frontend-ssr/redux/user/user.reducer.ts

```diff
@@ -0,0 +1,84 @@
+import { UserState } from './types';
+import { USER_NOTIFICATIONS } from './user.actions';
+
+const initialState: UserState = {
+  info: undefined,
+  notifications: { list: undefined, unread: 0, unreadCount: 0 },
+  page: 0,
+  end: false,
+};
+
+export const userReducer = (state = initialState, action: { type: string; payload: any; })
+  : UserState => {
+  switch (action.type) {
+    case USER_NOTIFICATIONS.ADD: {
+      let { index, notification } = action.payload;
+
+      if (!Array.isArray(notification)) {
+        notification = [notification];
+      }
+      return {
+        ...state,
+        notifications: {
+          ...state.notifications,
+          list: [...state.notifications.list.slice(0, index),
+            ...notification,
+            ...state.notifications.list.slice(index)],
+        },
+      };
+    }
+
+    case USER_NOTIFICATIONS.LOAD: {
+      const { newNotifications } = action.payload;
+
+      return {
+        ...state,
+        notifications: {
+          ...state.notifications,
+          list: state.notifications.list
+            ? [...state.notifications.list, ...newNotifications]
+            : newNotifications,
+        },
+        page: state.page + 1,
+        end: newNotifications.length === 0,
+      };
+    }
+    case USER_NOTIFICATIONS.MARK_AS_READ: {
+      let { id } = action.payload;
+
+      return {
+        ...state,
+        notifications: {
+          ...state.notifications,
+          list: [{ ...state.notifications.list.find((n) => n._id === id), isUnread: false }],
+          unreadCount: state.notifications.unreadCount - 1,
+        },
+      };
+    }
+    case USER_NOTIFICATIONS.MARK_AS_UNREAD: {
+      let { id } = action.payload;
+
+      return {
+        ...state,
+        notifications: {
+          ...state.notifications,
+          list: [{ ...state.notifications.list.find((n) => n._id === id), isUnread: true }],
```

Comment:
Same here. You should find a way to modify just one notification, while keeping the rest.